src/tasks.ts

```diff
@@ -0,0 +1,63 @@
+import { Either, isLeft } from "fp-ts/lib/Either";
+import { Task } from "fp-ts/lib/Task";
+import { Millisecond } from "./units";
+
+export type TransientError = "transient";
+export const TransientError: TransientError = "transient";
+
+export type MaxRetries = "max-retries";
+export const MaxRetries: MaxRetries = "max-retries";
+
+/**
+ * A Task that can be retried when it fails with a transient error
+ */
+export type RetriableTask<E, T> = E extends TransientError
+  ? never
+  : Task<Either<E | TransientError, T>>;
+
+/**
+ * Wraps a RetriableTask with a number of retries
+ */
+export function withRetries<E, T>(
+  maxRetries: number,
+  backoff: (count: number) => Millisecond,
+  task: RetriableTask<E, T>
+): Task<Either<E | MaxRetries, T>> {
+  const runTaskOnce = (
+    count: number,
+    currentTask: RetriableTask<E, T>
+  ): Task<Either<E | MaxRetries, T>> => {
+    // on first execution, count === 0
+    if (count >= maxRetries - 1) {
+      // no more retries left
+      return currentTask.map(e =>
+        // if the task fails with a TransientError, we map the error to
+        // a MaxRetries error
+        e.mapLeft(l => (l === TransientError ? MaxRetries : l))
+      );
+    }
+    return currentTask.chain(r => {
+      if (isLeft(r) && r.value === TransientError) {
```

Comment:
it's not clear, by the tipy definition it looks like task can return a left of only TransientErrors (E extends TransientError)
`Task<Either<E | TransientError, T>>`. what if we want to return a PermanentError from the task to stop retrying ?

Reply:
actually it's the opposite `E` can never extends from `TransientError`, this is to make sure that the `Task` can resolve to an error that is not `TransientError`:

```ts
type RetriableTask<E, T> = E extends TransientError
  ? never
  : Task<Either<E | TransientError, T>>;
```

Reply:
now i get it